### DISCUSSION TITLE
Fine tune Detection and Recognition models on custom dataset

### DISCUSSION THREAD

**Discussion by hkhanr** [mindee/doctr#637] [Q&A] [answered]
Hello,

I want to fine-tune detection and recognition models on my custom dataset. 
I have annotated few document images. I converted these annotation to the format mention in the following link [https://github.com/mindee/doctr/tree/main/references/detection](url) and finetuned a detection model. Now I want to finetune the recognition model. However, the given link doesn't clearly tells what should be the annotation format [https://github.com/mindee/doctr/tree/main/references/recognition](url). I mean the given link doesn't tells how to associate the GT box/polygon with GT text. 
My question is, how to associate GT polygon with GT text in polygon while training recognition model?

**Comment by fg-mindee** [comment, marked as answer]
Hi @hkhanr :wave: 

About the recognition format, I'm not sure I understand your question, but the [README](https://github.com/mindee/doctr/blob/main/references/recognition/README.md) does provide examples about the format. But text recognition doesn't have any localization information. Text detection does.

So if you have the OCR annotations (localization + text) of some documents. For the text detection training, only remove the text information. For the text recognition, you will have to crop each localized word. 

And then, imagine you have a folder called "images" and put the cropped images in it: `crop1.jpg, crop2.jpg, crop3.jpg`. Then your labels should be in json format in a "labels.json" next to the "images" folder, and should have this structure:
```
{"crop1.jpg": "I", "crop2.jpg": "love", "crop3.jpg": "OCR"}

```

Otherwise, end-to-end OCR training is not available in docTR :) I hope that helps!

**Reply by hkhanr** [reply]
Hello @fg-mindee , thanks for answering. Your answer has helped me a lot. In addition to this, I have another question. 
While training detection model, the detector expects bounding box in the form of polygon (which might be a skewed or rotated box) while the output is uniform rectangular box (xmin,ymin,xmax,ymax).  Why is this so? The detector doesn't detects the oriented box/text?

**Reply by fg-mindee** [reply]
Oh, that indeed needs some clarifications from our end:
- historically, our annotations are polygons so that, as you mentioned, we can have skewed boxes as well as straight boxes. And it's easy to convert your straight box annotations into a polygons of 4 points. We'll improve the dataset interface in the coming releases :+1: 
- on the last stable release, this isn't enabled by default but it you install docTR from source, this will be the case: the detection model can have multiple behaviours depending on your use case, which you can change with the `assume_straight_pages` & `export_as_straight_boxes` [flags](https://mindee.github.io/doctr/latest/models.html#doctr.models.ocr_predictor), that made their way in the high-level API in #632. The first flag controls whether the model assumes that all textual elements are horizontal, and the second one is about whether it should convert predictions to enclosing straight boxes if the previous flag is set to False.

I hope this helps :)

**Reply by hkhanr** [reply]
Great, so it means I can train model on skewed boxes. So, if export_as_straight_boxes = False, the model return detected polygons?

What if we want train recognition model on skewed boxes? As you said earlier that the recognition models is trained on cropped boxes. In that case if we have rotated/skewed boxes, how will we crop them? I mean we can only crop straight boxes. Do we have to crop straight boxes out of rotated boxes?

**Reply by fg-mindee** [reply]
> Great, so it means I can train model on skewed boxes. So, if export_as_straight_boxes = False, the model return detected polygons?

Yup!

> What if we want train recognition model on skewed boxes? As you said earlier that the recognition models is trained on cropped boxes. In that case if we have rotated/skewed boxes, how will we crop them? I mean we can only crop straight boxes. Do we have to crop straight boxes out of rotated boxes?

It wraps & rotated the boxes so that the recognition module sees almost straight crops :+1: We do crop rotated boxes with some warping!
In our next iteration we'll have a bit more rotation to the text recognition trainings!

**Reply by hkhanr** [reply]
So, for training recognition model, we manually have to unwarp and rotate the boxes so that they becomes straight, right?  Well, if this is the case, can you share the piece of code for cropping straight boxes out of rotated polygons for training recognition model?

**Reply by fg-mindee** [reply]
Of course, I suggest checking how we do this in the predictors: https://github.com/mindee/doctr/blob/main/doctr/models/predictor/pytorch.py#L64-L67

which uses this method: https://github.com/mindee/doctr/blob/main/doctr/models/predictor/base.py#L49-L67

Let me know if things are unclear (since those are private methods, it's not documented as well as public methods/functions :sweat_smile:)

**Reply by hkhanr** [reply]
@fg-mindee  Thanks for sharing this method. Btw, I have written my own code of applying perspective transform on vertices of polygons and converted them into (xmin, ymin, xmax, ymax).
